### PR TITLE
Use SafeHandle for BPF Program

### LINKED ISSUE #311: pcap_compile() in Libpcap pre 1.8.0 is not thread-safe
Currently sharppcap provides static methods for creating/running filter, such as `PcapDevice.CheckFilter`.

Those methods use `pcap_compile`, this method is not thread-safe in old versions of Libpcap.

Having a static method that is not thread safe, could lead to crashes for multi-threaded applications.

Test that currently detect this issue is `MemorySafetyTests.DisposeDuringCapture`

### PR DESCRIPTION
**Pull request by kayoub5**
Fixes #311

### CODE REVIEW

**Comment by chmorgan** [comment]
@kayoub5 I saw your msg about pushing a new sharppcap rev. Would it make sense to do that after this PR is merged in? I can do it once it's in.

I rebuilt on circleci and seeing two tests fail rather than one for the previous build, hope the circleci logs provide some info.

**Comment by kayoub5** [comment]
@chmorgan without a crash dump, not much for me to base on to debug.

To generate a crash dump, we need to upgrade the entire CI to use .NET 5 instead of .NET 3.1 (`dotnet test --blame-crash` available since .NET 5.0 preview SDK)

**Comment by chmorgan** [comment]
doh. @kayoub5 . I tried a branch with dotnet-sdk-5.0 but this results in runtime/build errors due to lack of finding a compatible version (3.1). Thoughts on how we should proceed? https://app.circleci.com/pipelines/github/chmorgan/sharppcap/686/workflows/dd01fab6-8acd-466f-90c8-7000036e492e/jobs/2425

**Comment by kayoub5** [comment]
> 
> 
> doh. @kayoub5 . I tried a branch with dotnet-sdk-5.0 but this results in runtime/build errors due to lack of finding a compatible version (3.1). Thoughts on how we should proceed? https://app.circleci.com/pipelines/github/chmorgan/sharppcap/686/workflows/dd01fab6-8acd-466f-90c8-7000036e492e/jobs/2425

Did you change dotnet version in https://github.com/chmorgan/sharppcap/blob/master/scripts/install-dotnet.sh ?

**Comment by chmorgan** [comment]
That’s where I changed it yep, but on the branch.

**Comment by kayoub5** [comment]
@chmorgan 
* Good news: I manged to get stack trace of the crash
* Bad news: crash also happens in master see #310
* Good news: crash happens only in `pcap_compile` when multiple threads try to compile a filter at the same time
* Bad news: we did not fix the crashes 100%, we just found new ones that are even more sporadic.

**Comment by chmorgan** [comment]
Hahah. That was an emotional roller coaster!

Look like this will help point to how to fix the corner case crashes?

I guess I should have used dotnet and not dotnet core with my edits to use
5.0?

**Comment by kayoub5** [comment]
Found it!
https://www.tcpdump.org/libpcap-changes.txt
````
    pcap_compile() in 1.8.0 and later is newly thread-safe.
````

**Comment by kayoub5** [comment]
> I guess I should have used dotnet and not dotnet core with my edits to use 5.0?

Yeah, in version 5 they just call it `dotnet`, not `dotnet-core`, also you need to update both the installation script and the `.csproj` files

**Comment by kayoub5** [comment]
@chmorgan we got two solutions:
* short term: disable tests on old Libpcap versions
* long term: Use a lock on pcap_compile function when using old Libpcap

Question is: do we even want to support such edge cases for old Libpcap?

**Comment by chmorgan** [comment]
Good thinking on considering whether we need to resolve the issue or not.

Ubuntu 18.04 is using 1.8.1 so it would seem we’d be safe to upgrade or
drop the old Ubuntu versions and libpcap versions at this point and add a
note to sharppcap readme.md mentioning the issue with pre 1.8 revs of
libpcap.

I can see about updating or dropping the old Ubuntu versions we are testing
with unless you wanted to take a look at that.

Chris

File: Test/MemorySafetryTests.cs [hunk 1]
```diff
@@ -25,7 +25,9 @@ public class MemorySafetryTests
 
         [Test]
         [Parallelizable(ParallelScope.Children)]
-        public void DisposeDuringCapture([Range(1, 10)] int _)
+        // pcap_compile() in 1.8.0 and later is newly thread-safe
```
Comment (chmorgan) [inline Test/MemorySafetryTests.cs:28]: what are the side effects here of not calling this routine? Are we not testing for behavior that we know will fail on these older libpcap revs? It's a little unclear what the intent is here. Also, filename should probably be MemorySafetyTests.cs (type in safety).

Comment (kayoub5) [reply Test/MemorySafetryTests.cs:28]: I am simply disabling this test on old Libpcap until we fix it, since it's causing false positives for all currently open PRs, 

The topic would be addressed in #311

Comment (chmorgan) [reply Test/MemorySafetryTests.cs:28]: @kayoub5 mind adding a comment that this is what you are doing? like:

//pcap_compile() in 1.8.0 and later is newly thread-safe
// disable dispose during capture tests in earlier versions to avoid crashes due to known lack of thread-safety

Comment (kayoub5) [reply Test/MemorySafetryTests.cs:28]: I moved those changes to #312, could we have that one merged first?

**Review by chmorgan — APPROVED** [review]

**Comment by kayoub5** [comment]
@chmorgan last review, #311 is now fixed as well

**Review by chmorgan — APPROVED** [review]
